### PR TITLE
Add a snippet system

### PR DESCRIPTION
**Pull request by pascalkuthe**
This PR finally implements a real snippet system that can scale Fo fully support all features that you expect from snippets. It has been moved to helix core to reflect that these snippets can be sourced from other sources besides LSPs in the future.

I kept the snippet parsers and some of the logic around generating transactions but basically had to fully rewrite the snippet rendering since there were a ton of cases that weren't covered. Our old code around handling snippet indentation was also needlessly complicated, inefficient and incorrect which I fixed here too.

The actual tabstop handling (and its interaction with multi cursor) turned out to be quite tricky too. I created the notion of an active snippet which tracks the currently active tabstops on a document and maps them trough changes as needed. A snippet stays valid as long as all multi cursors are within the entire body of any snippet. I always hated how some vim snippet plugins made snippets way too sticky (so I would jump somewhere else, do something else hit tab and be back at the other end of a file at a renegade tabstop). This approach prevents that while still allowing you to enter normal mode and perform other edits *on the snippet* without immidietly discarding tabstops.

One you jump to a tabstop the placeholder will be selected, if you are in insertmode hitting any key not bound to a command will delete the placeholder text before inserting the corresponding character. This reuses (and expanded version) of our `on_next_key` mechanism.

Commits are best reviewed separately otherwise the diff may be hard to follow.

### CODE REVIEW

**Comment by pascalkuthe** [comment]
I am not yet sure I am happy with the keybindings, there is no way to jump back to a previous tabstop, the interaction with smart tab is hardcoded and normal mode bindings are missing. All of this somewhat conflicts with the jump to parent node end stuff. What are your thoughts @dead10ck ?

**Comment by dead10ck** [comment]
Hard coding the interaction with smart tab doesn't seem wrong on the face of it. The desired behavior is for tab to do different things in different situations, and this seems like a new situation.

That said, that doesn't solve going back. This kind of seems like a feature that deserves its own special mode, like view mode. In this mode, we could have tab / shift-tab in addition to a set of duplicate bindings for non-kitty protocol terminals. And it would also help avoid having to write a `smart-untab` command.

**Comment by pascalkuthe** [comment]
I thought we couldn't bins-tab at all? Or is that only because of the conflict with c-i? In insertmode c-i shouldn't be bound either so if we can create shift-tab bindings then that may work too..

One concern I had about hardworking tab is that I sometimes overshoot my target with tabstops and then endup doing a pretty for jump with the TS motion. So I guess I am not too much of a fan of having multiple things bound to the same key. Maybe that's just me but I guess having some configurability for smart tab could be nice.

**Comment by archseer** [comment]
When I started on this I figured snippets could be a separate sticky mode similar to how the debugger functions. An overlay on top of the regular keymap. Tabbing through all the tabstops would exit the mode too

**Comment by dead10ck** [comment]
Yeah, exactly my thought too. If it's a special mode / overlay, then we don't have to worry about conflicts or interacting with smart tab.

Also @pascalkuthe tab does conflict with Ctrl-i, so that couldn't be a default binding in normal mode. We'll have to figure out a different keymap for non-kitty terminals anyway though. Mod-tab only works in kitty terminals.

**Comment by dead10ck** [comment]
We could save the special mode for a follow-up PR though, and for now just come up with the bindings for non-kitty terminals. It will just probably not be the tab key.

**Comment by pascalkuthe** [comment]
yeah I think having it be mode based would be nice. It's entirely possible to bind `<S-tab>` in insert mode it just bound to `insert_tab` by default which would be changed by the mode. 

I think minor modes need some love in general. That would be a larger change tough so I think its best to just push that to follow-up work and only keep the very basic smart tab binding we have right now

File: helix-core/src/case_conversion.rs
Comment (matoous) [inline helix-core/src/case_conversion.rs:1]: **question**: (unrelated to the PR) if we will already have these inside helix, would it eventually make sense to offer commands for transforming strings before different casing?

File: helix-core/src/snippets/elaborate.rs [hunk 1]
```diff
@@ -0,0 +1,370 @@
+use std::mem::swap;
+use std::ops::Index;
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
```
Comment (the-mikedavis) [inline helix-core/src/snippets/elaborate.rs:5]: What do you think about avoiding anyhow in `helix-core`? To me anyhow seems useful for stuff like helix-term and commands but for library-like stuff like helix-core it makes it easy to not introduce custom Error types. Since we only have one usage here it would be easy to drop and replace with something thin like a struct + thiserror

File: helix-core/src/snippets/elaborate.rs [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+            }
+        };
+        Some(Transform {
+            regex,
+            global,
+            replacement: transform.replacement.into(),
+        })
+    }
+
+    // TODO: use regex cursor so we can use a rope slice
```
Comment (the-mikedavis) [inline helix-core/src/snippets/elaborate.rs:308]: This change predates regex-cursor being merged into master, right? Could this be updated now to use regex-cursor?

File: helix-core/src/snippets/elaborate.rs (hunk 1 above)
Comment (pascalkuthe) [reply helix-core/src/snippets/elaborate.rs:5]: Sounds good

File: helix-stdx/src/range.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/// A range of `char`s within the text.
+#[derive(Debug, Clone, Copy, PartialOrd, Ord, PartialEq, Eq)]
+pub struct Range<T = usize> {
+    pub start: T,
+    pub end: T,
+}
+
+impl<T: PartialOrd> Range<T> {
+    pub fn contains(&self, other: Self) -> bool {
+        (self.start <= other.start) & (other.end <= self.end)
```
Comment (the-mikedavis) [inline helix-stdx/src/range.rs:10]:
```suggestion
        (self.start <= other.start) && (other.end <= self.end)
```

File: helix-core/src/case_conversion.rs
Comment (erasebegin) [reply helix-core/src/case_conversion.rs:1]: Maybe I'm misunderstanding but Helix already has a way to transform the selected text to uppercase. Maybe lowercase too, I'm not sure.

**Comment by nour-dev** [comment]
Can I use this feature if I build helix from source ?

**Comment by zetashift** [comment]
> Can I use this feature if I build helix from source ?

If you build from the right branch yes!
This is the branch you want to checkout: https://github.com/helix-editor/helix/tree/snippet_placeholder

And if you're a `nix` user you can try it out in a single command:
`nix profile install github:helix-editor/helix#snippet_placeholder`

**Comment by nour-dev** [comment]
Is there any docs about making a snippet ?
Or is it simple like vs code snippets ? In json format ?

**Comment by pascalkuthe** [comment]
Ots vscode snippets but this doesn't handle leading snippets, it only Prozesses snippets send by the lsp

**Comment by nour-dev** [comment]
> it's vscode snippets but this doesn't handle loading user snippets, it only handles snippets send by the lsp

For now I am using [snippets lsp](https://github.com/estin/simple-completion-language-server). But the stop points are not working. So if I build helix with this branch . The lsp will work with stop points ?

**Comment by pascalkuthe** [comment]
Yes

**Comment by danillos** [comment]
@pascalkuthe I was testing and tabstop wasn't working. So I realized that it only works when `smart-tab` is enabled.

Shouldn't it work even when smart-tag disabled? Since they are different features.

**Comment by pascalkuthe** [comment]
More advanced keybindings are left to future work. You can bind the commands to jump to the previous/next tabstop commands manually to different keya

**Comment by nour-dev** [comment]
Started to test it today. Generally, it is working very well with the mentioned lsp before.

Thank you very much, @pascalkuthe

**Comment by DaniD3v** [comment]
> And if you're a `nix` user you can try it out in a single command:
> `nix profile install github:helix-editor/helix#snippet_placeholder`

Just wanted to mention that it's `nix profile install github:helix-editor/helix/snippet_placeholder`
or the shell version: `nix shell github:helix-editor/helix/snippet_placeholder`

"#" selects a flake attribute in nix, not a git branch

**Comment by GameDungeon** [comment]
This might be out of scope for this PR (and if it is it might be good to open another issue or add a comment on the existing one), but I notice that this PR implements parsing for variables but does not resolve any of the special ones I see in the spec, e.g. TM_SELECTED_TEXT This would be a very nice addition that seems fairly easy to add on top of the amazing work in this PR.

Otherwise, I've been using this branch for a few days and the snippet support it provides is an amazing upgrade with no obvious bugs.

**Comment by andreytkachenko** [comment]
I have been using this feature for couple months and it works perfect. Thank you @pascalkuthe! Is anything blocking that to be in master?

**Comment by schlich** [comment]
more messing around in nix; here's a `flake.nix` that at least built for me using nix-direnv  (swap out the `aarch64-darwin` if you're not on mac silicon)

```
{
  description = "Flake for source install of helix with snippets.";
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs?ref=nixos-unstable";
    helix.url = "github:helix-editor/helix?ref=snippet_placeholder";
  };
  outputs = { self, nixpkgs, helix }:
    let
      helix-dev = helix.packages.aarch64-darwin;
    in
    {
      packages.aarch64-darwin.helix = helix.packages.aarch64-darwin.helix;
      devShells.aarch64-darwin.default = nixpkgs.legacyPackages.aarch64-darwin.mkShell{
        buildInputs = [ helix-dev.helix ];
      };
    };
}
```

Now, for the life of me I can't figure out how to try this out myself.  Is there a recommended language/lsp config for testing this out? Will docs about this change be incorporated?  Willing to help with that if possible...

**Comment by GameDungeon** [comment]
Personally I have been using this branch with this https://github.com/estin/simple-completion-language-server lsp that has a flake. Testing with that, the tabstops and everything else I try work great

**Comment by schlich** [comment]
here's an updated flake that loads both the patched version of helix and the snippets LSP using home manager:

(i'm new at nix so feedback welcome, esp. around naming conventions)

`~/.config/home-manager/flake.nix`

```nix
{
  description = "Home Manager configuration of schlich";

  inputs = {
    # Specify the source of Home Manager and Nixpkgs.
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    home-manager = {
      url = "github:nix-community/home-manager";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    helix.url = "github:helix-editor/helix?ref=snippet_placeholder";
    scls.url = "github:estin/simple-completion-language-server";
    
  };

  outputs = { nixpkgs, home-manager, helix, scls, ... }:
    let
      system = "aarch64-darwin";
      pkgs = nixpkgs.legacyPackages.${system};
      helix-dev = helix.packages.${system};
      scls-dev = scls.defaultPackage.${system};
    in {
      homeConfigurations."schlich" = home-manager.lib.homeManagerConfiguration {
        inherit pkgs;

        # Specify your home configuration modules here, for example,
        # the path to your home.nix.
        modules = [ ./home.nix ];

        # Optionally use extraSpecialArgs
        # to pass through arguments to home.nix
        extraSpecialArgs = { helixpkg = helix-dev; sclspkg = scls-dev; };
      };
    };
}
```

`~/.config/home-manager/home.nix`

```nix
{ pkgs, helixpkg, sclspkg, ... }:

{
  ...
  home.packages = [
    helixpkg.helix
    sclspkg
  ];
  ...
}

relevant documentation for [smart-tab](https://docs.helix-editor.com/editor.html?highlight=smart-tab#editorsmart-tab-section)

are y'all enabling the `supercede-menu` option or something else?  Should we be going with the keybindings suggested in that section?

File: helix-core/src/case_conversion.rs
Comment (rcorre) [reply helix-core/src/case_conversion.rs:1]: For reference, there is an open issue about this: https://github.com/helix-editor/helix/issues/5197. Helix does have lower<->upper conversion, but not snake/pascal/etc.

**Comment by UltraBlackLinux** [comment]
any updates on this? No new commits in months, no new comments in weeks - What is holding this PR back? A Lack of activity? Are there any issues or uncertainties?

**Comment by Philipp-M** [comment]
FWIW, I merged current master (after #2608 was merged) into this branch here: https://github.com/Philipp-M/helix/tree/snippet_placeholder
I have tried to isolate the changes that are relevant to solve the merge conflicts here: https://github.com/Philipp-M/helix/commit/18c0d36b6afaa36bf20924eba554e940721454e7
The diff is a little bit annoying, as not much happened semantically, but since code was moved and git hasn't recognized this, it's larger than justified.

I'm not sure if I find time for a more in-depth review, but after a skim it looks solid (as usual).
As I think it was tested rather exhaustively by various people (and includes quite a few tests), maybe merge this to increase testing-surface and iterate on it in follow-up PRs (if necessary)?

**Comment by the-mikedavis** [comment]
I want to give this another look in the next few days. I've been using it locally for a while but I've seen some bugs where the tabstop highlight doesn't show up or stick around consistently. (It's possible that I just messed up a merge in my branch though - Pascal was unable to reproduce what I was seeing.) I'll give it a rebase when I give it another look.

There was some discussion above about maybe using a minor mode instead of some of the new code for handling the tab keypress while within a snippet that I'd like to consider again. Chatting with @pascalkuthe it sounds like this may be possible but might need more changes to the keymap module.

**Review by the-mikedavis — APPROVED** [review]
I took another look through this and it looks great! The main part that I was worried about - the extra keymap handling code for tabstops - is very minimal with this PR. I'll add an issue to replace that as we refactor minor modes (as mentioned above).

**Comment by filipdutescu** [comment]
Amazing work done here! Congratulations to everyone involved, I genuinely think this will be an incredibly useful feature and will improve quality of life by a lot! Very excited to try it out!

**Comment by dpc** [comment]
With a bar cursor-shape, switching to next snippet tabstop displays kind of weirdly...

The first tabstop is OK:

image

but after typing the text and using `tab` to jump to the next one it shows like this:

image

same for all the following ones.

**Comment by dpc** [comment]
Also - with inability to fully customize tab handling here, a new keybinding that would close the completion without changing the mode would be great. Then if I'd want to jump to next tabstop while the completion is opened, I would just close it and press `tab`. Currently pressing escape closes it, but also quits inserting / snippet mode altogether. It seems to me that a new command to just quit completion here: https://docs.helix-editor.com/keymap.html#completion-menu would be easy to implement.

**Comment by TornaxO7** [comment]
> With a bar cursor-shape, switching to next snippet tabstop displays kind of weirdly...
> 
> The first tabstop is OK:
> 
> image
> 
> but after typing the text and using `tab` to jump to the next one it shows like this:
> 
> image
> 
> same for all the following ones.

I'd create a new issue for that

**Comment by gabydd** [comment]
Bit confused by this, you can just press enter that will accept the completion and close the menu while keeping the placeholders

> Also - with inability to fully customize tab handling here, a new keybinding that would close the completion without changing the mode would be great. Then if I'd want to jump to next tabstop while the completion is opened, I would just close it and press `tab`. Currently pressing escape closes it, but also quits inserting / snippet mode altogether. It seems to me that a new command to just quit completion here: https://docs.helix-editor.com/keymap.html#completion-menu would be easy to implement.

**Comment by gabydd** [comment]
Also there is already an issue https://github.com/helix-editor/helix/issues/1977 which I think should cover this behaviour

**Comment by dpc** [comment]
> Bit confused by this, you can just press enter that will accept the completion and close the menu while keeping the placeholders

I don't want to accept the completion though. E.g. I typed something, completions showed up, I decided I can't fill this tabstop ATM, so I want to jump to the next one.

> Also there is already an issue https://github.com/helix-editor/helix/issues/1977 which I think should cover this behaviour

It does seem inconsistent that the first tabstop is selected differently than other ones.

Correction: after trying it couple more times, it seems like the exact direction of the tabstop selection is just not reset, so sometimes it's forwards and sometimes backwards, probably just the last direction that was there.

**Comment by Iorvethe** [comment]
For closing the completion menu, it seems that `Control-C` is what you’re looking for
